src/base/torrentfileswatcher.cpp

```diff
@@ -0,0 +1,618 @@
+    {
+        QJsonObject jsonObj {
+            {PARAM_CATEGORY, params.category},
+            {PARAM_TAGS, serializeTagSet(params.tags)},
+            {PARAM_SAVEPATH, params.savePath},
+            {PARAM_OPERATINGMODE, Utils::String::fromEnum(params.addForced
+                ? BitTorrent::TorrentOperatingMode::Forced : BitTorrent::TorrentOperatingMode::AutoManaged)},
+            {PARAM_UPLOADLIMIT, params.uploadLimit},
+            {PARAM_DOWNLOADLIMIT, params.downloadLimit},
+            {PARAM_SEEDINGTIMELIMIT, params.seedingTimeLimit},
+            {PARAM_RATIOLIMIT, params.ratioLimit}
+        };
+
+        if (params.addPaused)
+            jsonObj[PARAM_STOPPED] = *params.addPaused;
+        if (params.contentLayout)
+            jsonObj[PARAM_CONTENTLAYOUT] = Utils::String::fromEnum(*params.contentLayout);
+        if (params.useAutoTMM)
+            jsonObj[PARAM_AUTOTMM] = *params.useAutoTMM;
+
+        return jsonObj;
+    }
+
+    TorrentFilesWatcher::WatchedFolderOptions parseWatchedFolderOptions(const QJsonObject &jsonObj)
+    {
+        TorrentFilesWatcher::WatchedFolderOptions options;
+        options.addTorrentParams = parseAddTorrentParams(jsonObj.value(OPTION_ADDTORRENTPARAMS).toObject());
+        options.recursive = jsonObj.value(OPTION_RECURSIVE).toBool();
+
+        return options;
+    }
+
+    QJsonObject serializeWatchedFolderOptions(const TorrentFilesWatcher::WatchedFolderOptions &options)
+    {
+        return {
+            {OPTION_ADDTORRENTPARAMS, serializeAddTorrentParams(options.addTorrentParams)},
+            {OPTION_RECURSIVE, options.recursive}
+        };
+    }
+}
+
+const int addTorrentParamsId = qRegisterMetaType<BitTorrent::AddTorrentParams>();
+const int magnetUriId = qRegisterMetaType<BitTorrent::MagnetUri>();
+const int torrentInfoId = qRegisterMetaType<BitTorrent::TorrentInfo>();
+
+TorrentFilesWatcher *TorrentFilesWatcher::m_instance = nullptr;
+
+void TorrentFilesWatcher::initInstance()
+{
+    if (!m_instance)
+        m_instance = new TorrentFilesWatcher;
+}
+
+void TorrentFilesWatcher::freeInstance()
+{
+    delete m_instance;
+    m_instance = nullptr;
+}
+
+TorrentFilesWatcher *TorrentFilesWatcher::instance()
+{
+    return m_instance;
+}
+
+TorrentFilesWatcher::TorrentFilesWatcher(QObject *parent)
+    : QObject {parent}
+    , m_ioThread {new QThread(this)}
+    , m_asyncWorker {new TorrentFilesWatcherWorker}
+{
+    connect(m_asyncWorker, &TorrentFilesWatcherWorker::magnetFound, this, &TorrentFilesWatcher::onMagnetFound);
+    connect(m_asyncWorker, &TorrentFilesWatcherWorker::torrentFound, this, &TorrentFilesWatcher::onTorrentFound);
+
+    m_asyncWorker->moveToThread(m_ioThread);
+    m_ioThread->start();
+
+    load();
+}
+
+TorrentFilesWatcher::~TorrentFilesWatcher()
+{
+    m_ioThread->quit();
+    m_ioThread->wait();
+    delete m_asyncWorker;
+}
+
+QString TorrentFilesWatcher::makeCleanPath(const QString &path)
```

Comment:
Shouldn't this be a free function?

Reply:
No. It has no generic meaning. It produces a path that is clean from perspective of `TorrentFilesWatcher` so I prefer to have it as class member (even though it doesn't use any private class members).

Reply:
Or this function is intended to be `const`?

Reply:
Or this function is intended to be `const`?

Reply:
>Or this function is intended to be `const`?

No. It is `static`.